Replace `IPvAnyAddress` and related types with `Union[IPv4Address, IPv6Address]`

A solid 100LOC that looks buggy and like it could be 3 lines of code:

https://github.com/pydantic/pydantic/blob/12fde9c8b270cc5e73b0cf8bce1fb2aff718fd70/pydantic/networks.py#L266-L376

Selected Assignee: @davidhewitt

Hi! I assume you are aware of this, but still want to point out that current implementation doesn't work well with type checkers and auto completion.

Consider this example

```python
import pydantic


class Test(pydantic.BaseModel):
    ip: pydantic.IPvAnyAddress


t = Test(ip=pydantic.IPvAnyAddress("127.0.0.1"))
# t.ip: ipaddress.IPv4Address
print(t.ip.version)
print(t.ip.is_private)
```

Running type cheker on that will result in

<details><summary>Mypy</summary>
<p>

```sh
tmp/pydantic_ip_any.py:12: error: "IPvAnyAddress" has no attribute "version"  [attr-defined]
tmp/pydantic_ip_any.py:13: error: "IPvAnyAddress" has no attribute "is_private"  [attr-defined]
Found 2 errors in 1 file (checked 1 source file)
```

</p>
</details> 

<details><summary>pyright</summary>
<p>

```sh
tmp/pydantic_ip_any.py
  tmp/pydantic_ip_any.py:10:13 - error: Argument of type "IPv4Address | IPv6Address" cannot be assigned to parameter "ip" of type "IPvAnyAddress" in function "__init__"
    Type "IPv4Address | IPv6Address" cannot be assigned to type "IPvAnyAddress"
      "IPv4Address" is incompatible with "IPvAnyAddress" (reportArgumentType)
  tmp/pydantic_ip_any.py:12:12 - error: Cannot access member "version" for type "IPvAnyAddress"
    Member "version" is unknown (reportAttributeAccessIssue)
  tmp/pydantic_ip_any.py:13:12 - error: Cannot access member "is_private" for type "IPvAnyAddress"
    Member "is_private" is unknown (reportAttributeAccessIssue)
3 errors, 0 warnings, 0 informations 
```

</p>
</details> 

And obviously no hints and lots of complaints from IDE.

Not sure what the ideal way to solve it would be, but maybe it's worth at least adding stubs that mirror `Ipaddress._IPAddressBase`?

--- a/pydantic/networks.py
+++ b/pydantic/networks.py
@@ -547,7 +547,7 @@ class NameEmail(_repr.Representation):
 
 
 IPvAnyAddressType: TypeAlias = 'IPv4Address | IPv6Address'
-IPvAnyInterfaceType: TypeAlias = 'IPv4Interface |IPv6Interface '
+IPvAnyInterfaceType: TypeAlias = 'IPv4Interface | IPv6Interface'
 IPvAnyNetworkType: TypeAlias = 'IPv4Network | IPv6Network'
 
 if TYPE_CHECKING:

--- a/pydantic/networks.py
+++ b/pydantic/networks.py
@@ -546,118 +546,120 @@ class NameEmail(_repr.Representation):
         return f'{self.name} <{self.email}>'
 
 
-class IPvAnyAddress:
-    """Validate an IPv4 or IPv6 address.
-
-    ```py
-    from pydantic import BaseModel
-    from pydantic.networks import IPvAnyAddress
-
-    class IpModel(BaseModel):
-        ip: IPvAnyAddress
-
-    print(IpModel(ip='127.0.0.1'))
-    #> ip=IPv4Address('127.0.0.1')
-
-    try:
-        IpModel(ip='http://www.example.com')
-    except ValueError as e:
-        print(e.errors())
-        '''
-        [
-            {
-                'type': 'ip_any_address',
-                'loc': ('ip',),
-                'msg': 'value is not a valid IPv4 or IPv6 address',
-                'input': 'http://www.example.com',
-            }
-        ]
-        '''
-    ```
-    """
-
-    __slots__ = ()
-
-    def __new__(cls, value: Any) -> IPv4Address | IPv6Address:
-        """Validate an IPv4 or IPv6 address."""
-        try:
-            return IPv4Address(value)
-        except ValueError:
-            pass
-
-        try:
-            return IPv6Address(value)
-        except ValueError:
-            raise PydanticCustomError('ip_any_address', 'value is not a valid IPv4 or IPv6 address')
-
-    @classmethod
-    def __get_pydantic_json_schema__(
-        cls, core_schema: core_schema.CoreSchema, handler: _schema_generation_shared.GetJsonSchemaHandler
-    ) -> JsonSchemaValue:
-        field_schema = {}
-        field_schema.update(type='string', format='ipvanyaddress')
-        return field_schema
-
-    @classmethod
-    def __get_pydantic_core_schema__(
-        cls,
-        _source: type[Any],
-        _handler: GetCoreSchemaHandler,
-    ) -> core_schema.CoreSchema:
-        return core_schema.no_info_plain_validator_function(
-            cls._validate, serialization=core_schema.to_string_ser_schema()
-        )
-
-    @classmethod
-    def _validate(cls, input_value: Any, /) -> IPv4Address | IPv6Address:
-        return cls(input_value)  # type: ignore[return-value]
-
-
-class IPvAnyInterface:
-    """Validate an IPv4 or IPv6 interface."""
-
-    __slots__ = ()
-
-    def __new__(cls, value: NetworkType) -> IPv4Interface | IPv6Interface:
-        """Validate an IPv4 or IPv6 interface."""
-        try:
-            return IPv4Interface(value)
-        except ValueError:
-            pass
-
-        try:
-            return IPv6Interface(value)
-        except ValueError:
-            raise PydanticCustomError('ip_any_interface', 'value is not a valid IPv4 or IPv6 interface')
-
-    @classmethod
-    def __get_pydantic_json_schema__(
-        cls, core_schema: core_schema.CoreSchema, handler: _schema_generation_shared.GetJsonSchemaHandler
-    ) -> JsonSchemaValue:
-        field_schema = {}
-        field_schema.update(type='string', format='ipvanyinterface')
-        return field_schema
-
-    @classmethod
-    def __get_pydantic_core_schema__(
-        cls,
-        _source: type[Any],
-        _handler: GetCoreSchemaHandler,
-    ) -> core_schema.CoreSchema:
-        return core_schema.no_info_plain_validator_function(
-            cls._validate, serialization=core_schema.to_string_ser_schema()
-        )
-
-    @classmethod
-    def _validate(cls, input_value: NetworkType, /) -> IPv4Interface | IPv6Interface:
-        return cls(input_value)  # type: ignore[return-value]
-
-
+IPvAnyAddressType: TypeAlias = 'IPv4Address | IPv6Address'
+IPvAnyInterfaceType: TypeAlias = 'IPv4Interface |IPv6Interface '
 IPvAnyNetworkType: TypeAlias = 'IPv4Network | IPv6Network'
 
 if TYPE_CHECKING:
+    IPvAnyAddress = IPvAnyAddressType
+    IPvAnyInterface = IPvAnyInterfaceType
     IPvAnyNetwork = IPvAnyNetworkType
 else:
+
+    class IPvAnyAddress:
+        """Validate an IPv4 or IPv6 address.
+
+        ```py
+        from pydantic import BaseModel
+        from pydantic.networks import IPvAnyAddress
+
+        class IpModel(BaseModel):
+            ip: IPvAnyAddress
+
+        print(IpModel(ip='127.0.0.1'))
+        #> ip=IPv4Address('127.0.0.1')
+
+        try:
+            IpModel(ip='http://www.example.com')
+        except ValueError as e:
+            print(e.errors())
+            '''
+            [
+                {
+                    'type': 'ip_any_address',
+                    'loc': ('ip',),
+                    'msg': 'value is not a valid IPv4 or IPv6 address',
+                    'input': 'http://www.example.com',
+                }
+            ]
+            '''
+        ```
+        """
+
+        __slots__ = ()
+
+        def __new__(cls, value: Any) -> IPvAnyAddressType:
+            """Validate an IPv4 or IPv6 address."""
+            try:
+                return IPv4Address(value)
+            except ValueError:
+                pass
+
+            try:
+                return IPv6Address(value)
+            except ValueError:
+                raise PydanticCustomError('ip_any_address', 'value is not a valid IPv4 or IPv6 address')
+
+        @classmethod
+        def __get_pydantic_json_schema__(
+            cls, core_schema: core_schema.CoreSchema, handler: _schema_generation_shared.GetJsonSchemaHandler
+        ) -> JsonSchemaValue:
+            field_schema = {}
+            field_schema.update(type='string', format='ipvanyaddress')
+            return field_schema
+
+        @classmethod
+        def __get_pydantic_core_schema__(
+            cls,
+            _source: type[Any],
+            _handler: GetCoreSchemaHandler,
+        ) -> core_schema.CoreSchema:
+            return core_schema.no_info_plain_validator_function(
+                cls._validate, serialization=core_schema.to_string_ser_schema()
+            )
+
+        @classmethod
+        def _validate(cls, input_value: Any, /) -> IPvAnyAddressType:
+            return cls(input_value)  # type: ignore[return-value]
+
+    class IPvAnyInterface:
+        """Validate an IPv4 or IPv6 interface."""
+
+        __slots__ = ()
+
+        def __new__(cls, value: NetworkType) -> IPvAnyInterfaceType:
+            """Validate an IPv4 or IPv6 interface."""
+            try:
+                return IPv4Interface(value)
+            except ValueError:
+                pass
+
+            try:
+                return IPv6Interface(value)
+            except ValueError:
+                raise PydanticCustomError('ip_any_interface', 'value is not a valid IPv4 or IPv6 interface')
+
+        @classmethod
+        def __get_pydantic_json_schema__(
+            cls, core_schema: core_schema.CoreSchema, handler: _schema_generation_shared.GetJsonSchemaHandler
+        ) -> JsonSchemaValue:
+            field_schema = {}
+            field_schema.update(type='string', format='ipvanyinterface')
+            return field_schema
+
+        @classmethod
+        def __get_pydantic_core_schema__(
+            cls,
+            _source: type[Any],
+            _handler: GetCoreSchemaHandler,
+        ) -> core_schema.CoreSchema:
+            return core_schema.no_info_plain_validator_function(
+                cls._validate, serialization=core_schema.to_string_ser_schema()
+            )
+
+        @classmethod
+        def _validate(cls, input_value: NetworkType, /) -> IPvAnyInterfaceType:
+            return cls(input_value)  # type: ignore[return-value]
 
     class IPvAnyNetwork:
         """Validate an IPv4 or IPv6 network."""

